File: 2341-maximum-number-of-pairs-in-array/2341-maximum-number-of-pairs-in-array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var numberOfPairs = function(nums) {
  const map = {};
  let pairCount = 0;
  let leftOver = 0;

  for (const num of nums) {
    map[num] = (map[num] || 0) + 1;
    if (map[num] % 2 === 0) {
      pairCount++;
    }
  }


  for (const key in map) {
    if (map[key] % 2 !== 0) {
      leftOver++;
    } 
  }
  return ([pairCount, leftOver]);
};

File: 2341-maximum-number-of-pairs-in-array/2341-maximum-number-of-pairs-in-array.py
class Solution:
    def numberOfPairs(self, nums: List[int]) -> List[int]:
        feq_dict = {}
        
        for num in nums:
            feq_dict[num] = feq_dict.get(num, 0) + 1
        
        count_pairs = 0
        count_left = 0
        
        for count in feq_dict.values():
            count_pairs += count // 2
            count_left += count % 2   
            
        return [count_pairs, count_left]